// data_api/dids_api.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * @author Pedro Sanders
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * @since v1
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      */


var _utils = require('core/utils');

var _utils2 = _interopRequireDefault(_utils);

var _utils3 = require('data_api/utils');

var _utils4 = _interopRequireDefault(_utils3);

var _status = require('core/status');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var DIDsAPI = function () {
    function DIDsAPI(dataSource) {
        _classCallCheck(this, DIDsAPI);

        this.ds = dataSource;
    }

    _createClass(DIDsAPI, [{
        key: 'createFromJSON',
        value: function createFromJSON(jsonObj) {
            var response = this.ds.withCollection('gateways').find("@.metadata.ref=='" + jsonObj.metadata.gwRef + "'");

            if (response.result.length == 0) {
                return _status.UNFULFILLED_DEPENDENCY_RESPONSE;
            }

            return this.didExist(jsonObj.spec.location.telUrl) ? _utils2.default.buildResponse(_status.Status.CONFLICT) : this.ds.insert(jsonObj);
        }
    }, {
        key: 'updateFromJSON',
        value: function updateFromJSON(jsonObj) {
            var response = this.ds.withCollection('gateways').find("@.metadata.ref=='" + jsonObj.metadata.gwRef + "'");

            if (response.result.length == 0) {
                return _status.UNFULFILLED_DEPENDENCY_RESPONSE;
            }

            return !this.didExist(jsonObj.spec.location.telUrl) ? _utils2.default.buildResponse(_status.Status.NOT_FOUND) : this.ds.update(jsonObj);
        }
    }, {
        key: 'getDIDs',
        value: function getDIDs(filter) {
            return this.ds.withCollection('dids').find(filter);
        }
    }, {
        key: 'getDID',
        value: function getDID(ref) {
            return _utils4.default.deepSearch(this.getDIDs(), "metadata.ref", ref);
        }

        /**
         * note: telUrl maybe a string in form of 'tel:${number}' or
         * a TelURL Object.
         */

    }, {
        key: 'getDIDByTelUrl',
        value: function getDIDByTelUrl(telUrl) {
            return _utils4.default.deepSearch(this.getDIDs(), "spec.location.telUrl", telUrl);
        }
    }, {
        key: 'didExist',
        value: function didExist(telUrl) {
            return _utils4.default.objExist(this.getDIDByTelUrl(telUrl));
        }
    }, {
        key: 'deleteDID',
        value: function deleteDID(ref) {
            return this.ds.withCollection('dids').remove(ref);
        }
    }]);

    return DIDsAPI;
}();

exports.default = DIDsAPI;
