// data_api/domains_api.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();
/**
 * @author Pedro Sanders
 * @since v1
 */


var _utils = require('core/utils');

var _utils2 = _interopRequireDefault(_utils);

var _utils3 = require('data_api/utils');

var _utils4 = _interopRequireDefault(_utils3);

var _status = require('core/status');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var foundDependentObjects = { status: _status.Status.CONFLICT, message: _status.Status.message[4092].value };

var DomainsAPI = function () {
    function DomainsAPI(dataSource) {
        _classCallCheck(this, DomainsAPI);

        this.ds = dataSource;
    }

    _createClass(DomainsAPI, [{
        key: 'createFromJSON',
        value: function createFromJSON(jsonObj) {
            return this.domainExist(jsonObj.spec.context.domainUri) ? _utils2.default.buildResponse(_status.Status.CONFLICT) : this.ds.insert(jsonObj);
        }
    }, {
        key: 'updateFromJSON',
        value: function updateFromJSON(jsonObj) {
            return !this.domainExist(jsonObj.spec.context.domainUri) ? _utils2.default.buildResponse(_status.Status.NOT_FOUND) : this.ds.update(jsonObj);
        }
    }, {
        key: 'getDomains',
        value: function getDomains(filter) {
            return this.ds.withCollection('domains').find(filter);
        }
    }, {
        key: 'getDomain',
        value: function getDomain(ref) {
            return _utils4.default.deepSearch(this.getDomains(), "metadata.ref", ref);
        }
    }, {
        key: 'getDomainByUri',
        value: function getDomainByUri(domainUri) {
            return _utils4.default.deepSearch(this.getDomains(), "spec.context.domainUri", domainUri);
        }
    }, {
        key: 'domainExist',
        value: function domainExist(domainUri) {
            return _utils4.default.objExist(this.getDomainByUri(domainUri));
        }
    }, {
        key: 'deleteDomain',
        value: function deleteDomain(ref) {
            var response = this.getDomain(ref);

            if (response.status != _status.Status.OK) {
                return response;
            }

            var domain = response.result;

            response = this.ds.withCollection('agents').find("'" + domain.spec.context.domainUri + "' in @.spec.domains");
            var agents = response.result;

            return agents.length == 0 ? this.ds.withCollection('domains').remove(ref) : foundDependentObjects;
        }
    }]);

    return DomainsAPI;
}();

exports.default = DomainsAPI;
