// data_api/gateways_api.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * @author Pedro Sanders
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * @since v1
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      */


var _utils = require('core/utils');

var _utils2 = _interopRequireDefault(_utils);

var _utils3 = require('data_api/utils');

var _utils4 = _interopRequireDefault(_utils3);

var _status = require('core/status');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var GatewaysAPI = function () {
    function GatewaysAPI(dataSource) {
        _classCallCheck(this, GatewaysAPI);

        this.ds = dataSource;
    }

    _createClass(GatewaysAPI, [{
        key: 'createFromJSON',
        value: function createFromJSON(jsonObj) {
            return this.gatewayExist(jsonObj.spec.host) ? _utils2.default.buildResponse(_status.Status.CONFLICT) : this.ds.insert(jsonObj);
        }
    }, {
        key: 'updateFromJSON',
        value: function updateFromJSON(jsonObj) {
            return !this.gatewayExist(jsonObj.spec.host) ? _utils2.default.buildResponse(_status.Status.NOT_FOUND) : this.ds.update(jsonObj);
        }
    }, {
        key: 'getGateways',
        value: function getGateways(filter) {
            return this.ds.withCollection('gateways').find(filter);
        }
    }, {
        key: 'getGateway',
        value: function getGateway(ref) {
            return _utils4.default.deepSearch(this.getGateways(), "metadata.ref", ref);
        }
    }, {
        key: 'getGatewayByHost',
        value: function getGatewayByHost(host) {
            return _utils4.default.deepSearch(this.getGateways(), "spec.host", host);
        }
    }, {
        key: 'gatewayExist',
        value: function gatewayExist(host) {
            return _utils4.default.objExist(this.getGatewayByHost(host));
        }
    }, {
        key: 'deleteGateway',
        value: function deleteGateway(ref) {
            var response = this.getGateway(ref);

            if (response.status != _status.Status.OK) {
                return response;
            }

            var gateway = response.result;

            response = this.ds.withCollection('dids').find("@.metadata.gwRef=='" + gateway.metadata.ref + "'");
            var dids = response.result;

            return dids.length == 0 ? this.ds.withCollection('gateways').remove(ref) : _status.FOUND_DEPENDENT_OBJECTS_RESPONSE;
        }
    }]);

    return GatewaysAPI;
}();

exports.default = GatewaysAPI;
