// data_api/peers_api.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * @author Pedro Sanders
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      * @since v1
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      */


var _utils = require('core/utils');

var _utils2 = _interopRequireDefault(_utils);

var _utils3 = require('data_api/utils');

var _utils4 = _interopRequireDefault(_utils3);

var _status = require('core/status');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var PeersAPI = function () {
    function PeersAPI(dataSource) {
        _classCallCheck(this, PeersAPI);

        this.ds = dataSource;
    }

    _createClass(PeersAPI, [{
        key: 'createFromJSON',
        value: function createFromJSON(jsonObj) {
            return this.peerExist(jsonObj.spec.credentials.username) ? _utils2.default.buildResponse(_status.Status.CONFLICT) : this.ds.insert(jsonObj);
        }
    }, {
        key: 'updateFromJSON',
        value: function updateFromJSON(jsonObj) {
            return !this.peerExist(jsonObj.spec.credentials.username) ? _utils2.default.buildResponse(_status.Status.NOT_FOUND) : this.ds.update(jsonObj);
        }
    }, {
        key: 'getPeers',
        value: function getPeers(filter) {
            return this.ds.withCollection('peers').find(filter);
        }
    }, {
        key: 'getPeer',
        value: function getPeer(ref) {
            return _utils4.default.deepSearch(this.getPeers(), "metadata.ref", ref);
        }
    }, {
        key: 'getPeerByUsername',
        value: function getPeerByUsername(username) {
            return _utils4.default.deepSearch(this.getPeers(), "spec.credentials.username", username);
        }
    }, {
        key: 'peerExist',
        value: function peerExist(username) {
            return _utils4.default.objExist(this.getPeerByUsername(username));
        }
    }, {
        key: 'deletePeer',
        value: function deletePeer(ref) {
            return this.ds.withCollection('peers').remove(ref);
        }
    }]);

    return PeersAPI;
}();

exports.default = PeersAPI;
